quantum/wpm.c

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2020 Richard Sutherland (rich@brickbots.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "wpm.h"
+
+//WPM Stuff
+static uint8_t current_wpm = 0;
+static uint8_t latest_wpm = 0;
+static uint16_t wpm_timer = 0;
+
+//This smoothing is 40 keystrokes
+static const float wpm_smoothing = 0.0487;
+
+void set_current_wpm(uint8_t new_wpm) { current_wpm = new_wpm; }
+
+uint8_t get_current_wpm(void) { return current_wpm;}
+
+void update_wpm(uint16_t keycode) {
+    if((keycode >= KC_A && keycode <= KC_0) || (keycode >= KC_TAB && keycode <= KC_SLASH)) {
```

Comment:
I think this block would be better if it was ... well handled like music mask.  

Eg, having a reasonable default is fine, but have it in a weakly defined function, so that users can overwrite which keycodes are okay.

Something like this: 
https://github.com/qmk/qmk_firmware/blob/1a79f14e4390d088fcf6b26fe3b62549875c3e71/quantum/process_keycode/process_music.c#L219-L235

That, or extract the keycode from modtap and layertap, and then check the keycode: 
https://github.com/qmk/qmk_firmware/blob/1a79f14e4390d088fcf6b26fe3b62549875c3e71/keyboards/crkbd/keymaps/drashna/keymap.c#L159-L163

Reply:
Seems like I could do both here.  Dig deeper into possible tap combos to make sure I don't miss any characters which should count towards WPM and then also move the logic for deciding to another function which can be overloaded.  

I'll make and test these changes and push them back to this branch.  Thank you for your time to review and for your suggestions!

Reply:
Well, if you want to do both, you could use the kb/user level function, and have it weakly set.  That ... would be the best of both, I think. :)

Reply:
Thanks @drashna!  This is what I ended up doing in the most recent commit.  I very much appreciated the pointers to the existing code.